src/mdns/mdns_mdnssd.cpp

```diff
@@ -703,6 +877,290 @@ void Publisher::Destroy(Publisher *aPublisher)
     delete static_cast<PublisherMDnsSd *>(aPublisher);
 }
 
+void PublisherMDnsSd::Subscription::Release(void)
+{
+    DeallocateServiceRef();
+}
+
+void PublisherMDnsSd::Subscription::DeallocateServiceRef(void)
+{
+    if (mServiceRef != nullptr)
+    {
+        DNSServiceRefDeallocate(mServiceRef);
+        mServiceRef = nullptr;
+    }
+}
+
+void PublisherMDnsSd::ServiceSubscription::Browse(void)
+{
+    assert(mServiceRef == nullptr);
+
+    otbrLog(OTBR_LOG_INFO, "[mdns] DNSServiceBrowse %s", mType.c_str());
+    DNSServiceBrowse(&mServiceRef, /* flags */ kDNSServiceFlagsTimeout, kDNSServiceInterfaceIndexAny, mType.c_str(),
+                     /* domain */ nullptr, HandleBrowseResult, this);
+}
+
+void PublisherMDnsSd::ServiceSubscription::HandleBrowseResult(DNSServiceRef       aServiceRef,
+                                                              DNSServiceFlags     aFlags,
+                                                              uint32_t            aInterfaceIndex,
+                                                              DNSServiceErrorType aErrorCode,
+                                                              const char *        aInstanceName,
+                                                              const char *        aType,
+                                                              const char *        aDomain,
+                                                              void *              aContext)
+{
+    static_cast<ServiceSubscription *>(aContext)->HandleBrowseResult(aServiceRef, aFlags, aInterfaceIndex, aErrorCode,
+                                                                     aInstanceName, aType, aDomain);
+}
+
+void PublisherMDnsSd::ServiceSubscription::HandleBrowseResult(DNSServiceRef       aServiceRef,
+                                                              DNSServiceFlags     aFlags,
+                                                              uint32_t            aInterfaceIndex,
+                                                              DNSServiceErrorType aErrorCode,
+                                                              const char *        aInstanceName,
+                                                              const char *        aType,
+                                                              const char *        aDomain)
+{
+    OTBR_UNUSED_VARIABLE(aServiceRef);
+
+    otbrLog(OTBR_LOG_INFO, "[mdns] DNSServiceBrowse reply: %s.%s%s inf %u, flags=%u, error=%d", aInstanceName, aType,
+            aDomain, aInterfaceIndex, aFlags, aErrorCode);
+
+    VerifyOrExit(aErrorCode == kDNSServiceErr_NoError);
+    VerifyOrExit(aFlags & kDNSServiceFlagsAdd);
+
+    DeallocateServiceRef();
+    Resolve(aInterfaceIndex, aInstanceName, aType, aDomain);
```

Comment:
I didn't put `DeallocateServiceRef()` inside `Resolve` because it's this `ServiceRef` is created by `DNSServiceBrowse`, so it's `DNSServiceBrowseReply`'s duty to clean the `ServiceRef` when it's not needed. 
So, we need to `DeallocateServiceRef()` in `HandleBrowseResult` (which is the DNSServiceBrowseReply), no matter `HandleBrowseResult` calls `Resolve` or not.

Reply:
Also makes sense to me. Thanks!